x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/vector/Knn.java

```diff
@@ -0,0 +1,292 @@
+
+public class Knn extends FullTextFunction implements OptionalArgument, VectorFunction {
+
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Knn", Knn::readFrom);
+
+    private final Expression field;
+    private final Expression options;
+
+    public static final Map<String, DataType> ALLOWED_OPTIONS = Map.ofEntries(
+        entry(K_FIELD.getPreferredName(), INTEGER),
+        entry(NUM_CANDS_FIELD.getPreferredName(), INTEGER),
+        entry(VECTOR_SIMILARITY_FIELD.getPreferredName(), FLOAT),
+        entry(BOOST_FIELD.getPreferredName(), FLOAT),
+        entry(KnnQuery.RESCORE_OVERSAMPLE_FIELD, FLOAT)
+    );
+
+    @FunctionInfo(
+        returnType = "boolean",
+        preview = true,
+        description = "Finds the k nearest vectors to a query vector, as measured by a similarity metric. "
+            + "knn function finds nearest vectors through approximate search on indexed dense_vectors.",
+        examples = {
+            @Example(file = "knn-function", tag = "knn-function"),
+            @Example(file = "knn-function", tag = "knn-function-options"), },
+        appliesTo = { @FunctionAppliesTo(lifeCycle = FunctionAppliesToLifecycle.DEVELOPMENT) }
+    )
+    public Knn(
+        Source source,
+        @Param(name = "field", type = { "dense_vector" }, description = "Field that the query will target.") Expression field,
+        @Param(
+            name = "query",
+            type = { "dense_vector" },
+            description = "Vector value to find top nearest neighbours for."
+        ) Expression query,
+        @MapParam(
+            name = "options",
+            params = {
+                @MapParam.MapParamEntry(
+                    name = "boost",
+                    type = "float",
+                    valueHint = { "2.5" },
+                    description = "Floating point number used to decrease or increase the relevance scores of the query."
+                        + "Defaults to 1.0."
+                ),
+                @MapParam.MapParamEntry(
+                    name = "k",
+                    type = "integer",
+                    valueHint = { "10" },
+                    description = "The number of nearest neighbors to return from each shard. "
+                        + "Elasticsearch collects k results from each shard, then merges them to find the global top results. "
+                        + "This value must be less than or equal to num_candidates. Defaults to 10."
+                ),
+                @MapParam.MapParamEntry(
+                    name = "num_candidates",
+                    type = "integer",
+                    valueHint = { "10" },
+                    description = "The number of nearest neighbor candidates to consider per shard while doing knn search. "
+                        + "Cannot exceed 10,000. Increasing num_candidates tends to improve the accuracy of the final results. "
```

Comment:
what happens if we do pass `num_candidates > 10000`? how do we surface this as an error?

Reply:
QueryBuilders get validated at construction time. Right now we surface QB construction errors as 400 with IAE:

```json
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_argument_exception",
                "reason": "[k] must be greater than 0"
            }
        ],
        "type": "illegal_argument_exception",
        "reason": "[k] must be greater than 0"
    },
    "status": 400
}
```